modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/js/categorization/categories/EditCategoriesModal.es.js

```diff
@@ -0,0 +1,365 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import ClayAlert from '@clayui/alert';
+import ClayButton from '@clayui/button';
+import {ClayRadio, ClayRadioGroup} from '@clayui/form';
+import ClayLoadingIndicator from '@clayui/loading-indicator';
+import ClayModal from '@clayui/modal';
+import {AssetCategoriesSelector} from 'asset-taglib';
+import {useIsMounted} from 'frontend-js-react-web';
+import {fetch} from 'frontend-js-web';
+import PropTypes from 'prop-types';
+import React, {useCallback, useContext, useEffect, useState} from 'react';
+
+import EditCategoriesContext from './EditCategoriesContext.es';
+
+const URL_SELECTION = '/bulk/v1.0/bulk-selection';
+const URL_UPDATE_CATEGORIES = '/bulk/v1.0/taxonomy-categories/batch';
+
+const EditCategoriesModal = ({
+	fileEntries = [],
+	folderId = '',
+	groupIds = [],
+	hiddenInput = 'assetCategoriesIds_',
+	observer,
+	onModalClose,
+	pathModule = '',
+	repositoryId = '',
+	selectAll = false,
+	selectCategoriesUrl = ''
+}) => {
+	const {namespace} = useContext(EditCategoriesContext);
+
+	const [append, setAppend] = useState(true);
+	const [description, setDescription] = useState('');
+
+	const [initialCategories, setInitialCategories] = useState([]);
+	// const [categories, setCategories] = useState([]);
+
+	const [isValid, setIsValid] = useState(true);
+
+	const [loading, setLoading] = useState(false);
+	const [multiple, setMultiple] = useState(false);
+
+	const [vocabularies, setVocabularies] = useState([]);
+	const [requiredVocabularies, setRequiredVocabularies] = useState([]);
+
+	const [selectedRadioGroupValue, setSelectedRadioGroupValue] = useState(
+		'add'
+	);
+
+	const bulkStatusComponent = Liferay.component(`${namespace}BulkStatus`);
+
+	const getDescription = size => {
+		if (size === 1) {
+			return Liferay.Language.get(
+				'you-are-editing-the-categories-for-the-selected-item'
+			);
+		}
+
+		return Liferay.Util.sub(
+			Liferay.Language.get(
+				'you-are-editing-the-common-categories-for-x-items.-select-edit-or-replace-current-categories'
+			),
+			size
+		);
+	};
+
+	const getFinalCategories = () => {
+		return vocabularies.reduce((accumulator, vocabulary) => {
+			vocabulary.selectedItems.map(({value}) => {
+				if (!accumulator.includes(value)) {
+					accumulator.push(parseInt(value, 10));
+				}
+			});
+			return accumulator;
+		}, []);
```

Comment:
Here the runtime may be as bad as cubic (`O(n^3)`). It seems the intent of the innermost loop is to `push` onto the `accumulator` only if the value isn't already in there: so you should be reaching for a `Set` straight away. You can turn it into an array at the end if that's what you need.

And `reduce` doesn't seem to be very useful here: you are always just mutating and returning the same `accumulator`, so the `reduce` call probably just makes things harder to read.

Something like this, for example, might address both of those issues:

```javascript
const categories = new Set();

vocabularies.forEach(({selectedItems}) => {
    selectedItems.map(({value}) => {
      categories.push(parseInt(value, 10));
    });
});

return Array.from(categories); // if you want this to be an array
```

Reply:
Much better thanks

Reply:
You probably meant `categories.add(parseInt(value, 10));` since it's a `Set` :grin: